pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+        d_den = self.rec(expr.denominator)
+        if len(d_den) > 1:
+            raise ValueError
+        den_k, den_v = list(d_den.items())[0]
+
+        result = {}
+        for k in d_num.keys():
+            result[d_num[k]/den_k] /= den_v
+        return result
+
+    def map_power(self, expr):
+        d_base = self.rec(expr.base)
+        d_exponent = self.rec(expr.exponent)
+        # d_exponent should look like {1: k}
+        if len(d_exponent) > 1 or 1 not in d_exponent:
+            raise RuntimeError("nonlinear expression")
+        exp = list(d_exponent.values())[0]
+        if exp == 1:
+            return d_base
+        if len(d_base) > 1:
+            raise NotImplementedError("powers are not implemented")
+        (k, v), = d_base.items()
+        return {k**exp: v**exp}
+
+    rec = __call__
+
+
+def _syzygy_module(m, generators):
+    """Takes as input a module of polynomials with domain :class:`sympy.EX`
+    represented as a matrix and returns the syzygy module as a matrix of polynomials
+    in the same domain. The syzygy module *S* that is returned as a matrix
+    satisfies S m = 0.
+    Using :class:`sympy.EX` because that represents the domain with any symbolic
+    element. Usually we need an Integer or Rational domain, but since there can be
+    unrelated symbols like *mu* in the expression, we need to use a symbolic domain.
+    """
+    from sympy.polys.orderings import grevlex
+
+    def _convert_to_matrix(module, *generators):
+        result = []
+        for syzygy in module:
+            row = []
+            for dmp in syzygy.data:
+                row.append(sympy.Poly(dmp.to_dict(), *generators,
+                    domain=sympy.EX).as_expr())
+            result.append(row)
+        return sympy.Matrix(result)
+
+    ring = sympy.EX.old_poly_ring(*generators, order=grevlex)
+    column_ideals = [ring.free_module(1).submodule(*m[:, i].tolist(), order=grevlex)
+                for i in range(m.shape[1])]
+    column_syzygy_modules = [ideal.syzygy_module() for ideal in column_ideals]
+
+    intersection = functools.reduce(lambda x, y: x.intersect(y),
+            column_syzygy_modules)
+
+    # _groebner_vec returns a groebner basis of the syzygy module as a list
+    groebner_vec = intersection._groebner_vec()
+    return _convert_to_matrix(groebner_vec, *generators)
+
+
+def _factor_left(mat, axis_vars):
+    """Return the left hand side of the factorisation of the matrix
+    For a matrix M, we want to find a factorisation such that M = L R
+    with minimum number of columns of L.
+    To get a good factorisation, what we do is first find a matrix
+    such that S M = 0 where S is the syzygy module converted to a matrix.
+    Then, M.T S.T = 0 which implies that M.T is in the space spanned by
+    the syzygy module of S.T and to get M we get the transpose of that.
+    """
+    first_syzygy_module = _syzygy_module(mat, axis_vars)
+    if len(first_syzygy_module) == 0:
```

Comment:
Add comment "left nullspace", basis in the rows.

Reply:
Done in 269fe81e798d6c048222baf0423cee65ed60ccee